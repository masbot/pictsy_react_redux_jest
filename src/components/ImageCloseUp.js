import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addComment } from '../actions/comments';

export class ImageCloseUp extends Component {

    constructor() {
        super();
        this.state = {
            comment: ''
        }
    }

    onInputChange = (event) => {
        this.setState({ comment: event.target.value });
    }

    addComment = (event) => {
        if(event.key == 'Enter') {
            const newComment = { id: this.props.image.id, text: this.state.comment};
            this.props.addComment(newComment);
            this.setState({comment: ''})
        }
    }

    renderComments() {
        if(!this.props.comments) {
            return <div></div>;
        }
        return this.props.comments.map( comment => {
            return <div className="comment-text" key={comment.id}>{comment.text}</div>
        })
    }

    render() {
        return (
            <div className="image-close-up">
                <div className="outer-image left-side">
                    <img src={`/public/assets/images/${this.props.image.url}`} alt="cat pic" onClick={() => this.props.onClickImage({})}/>
                </div>
                <div className="right-side">
                    <div className="comments">
                        {this.renderComments()}
                    </div>
                    <div>
                        <input 
                            className="comment-input" 
                            onChange={this.onInputChange} 
                            onKeyPress={this.addComment}
                            value={this.state.comment}
                            placeholder="Add Comment"
                            />
                    </div>
                </div>
            </div>
        )
    }
}

const mapStatetoProps = (state, ownProps) => {
    return {
        comments: state[ownProps.image.id]
    }
}

export default connect(mapStatetoProps, { addComment })(ImageCloseUp);