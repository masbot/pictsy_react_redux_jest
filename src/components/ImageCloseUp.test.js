import React from 'react';
import { shallow, mount } from 'enzyme';
import { ImageCloseUp } from './ImageCloseUp';

describe('ImageCloseUp', () => {
    const mockAddcomment = jest.fn();
    let props = { image: { url: 'location', id: "1"}, addComment: mockAddcomment };
    let imagecloseup = shallow(<ImageCloseUp {...props}/>);

    it('renders properly', () => {
        expect(imagecloseup).toMatchSnapshot();
    });

    it('creates input for comment input', () => {
        expect(imagecloseup.find('.comment-input').exists()).toBe(true);
    });

    it('initializes the `state` for comment', () => {
        expect(imagecloseup.state().comment).toEqual('');
    });

    describe('after receiving comments from redux', () => {
        const comments = {
            "1": [
                {
                    "id": "1",
                    "text": "1 comment A"
                },
                {
                    "id": "2",
                    "text": "1 comment B"
                },
                {
                    "id": "3",
                    "text": "1 comment C"
                }
            ]
        };

        beforeEach(() => {
            props = { image: { url: 'location', id: "1"}, comments };
            imagecloseup = shallow(<ImageCloseUp {...props}/>);
        });

        xit('creates comment element', () => {
            expect(imagecloseup.find('.comments').children().length).toEqual(comments[props.image.id].length);
        });
    });

    describe('when typing in comment input', () => {
        const comment = "first comment";

        beforeEach(() => {
            imagecloseup.find('.comment-input').simulate('change', { target: { value: comment}})
        });

        it('sets the `state` of comment to input value', () => {
            expect(imagecloseup.state().comment).toEqual(comment);
        });

        describe('when adding new comment', () => {
            beforeEach(( ) => {
                imagecloseup.find('.comment-input').simulate('keypress', { key: 'Enter'} );
            });
    
            it('dispatches addComment() from props', () => {
                expect(mockAddcomment).toHaveBeenCalled();
            });

            it('adds a new comment to comments list', () => {
                expect(imagecloseup.find('.comments').children.length).toEqual(1);
            });

            it('empties the input comment after pressing enter', () => {
                expect(imagecloseup.find('.comment-input').text()).toEqual('');
            })
        });
    });
});