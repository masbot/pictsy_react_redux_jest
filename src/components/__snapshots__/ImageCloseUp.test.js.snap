// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCloseUp renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImageCloseUp
    addComment={[Function]}
    image={
        Object {
            "id": "1",
            "url": "location",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="outer-image left-side"
>
          <img
                    alt="cat pic"
                    src="/public/assets/images/location"
          />
</div>,
        <div
          className="right-side"
>
          <div
                    className="comments"
          />
          <div>
                    <input
                              className="comment-input"
                              onChange={[Function]}
                              onKeyPress={[Function]}
                              value=""
                    />
          </div>
</div>,
      ],
      "className": "image-close-up",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt="cat pic"
            src="/public/assets/images/location"
/>,
          "className": "outer-image left-side",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "cat pic",
            "src": "/public/assets/images/location",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="comments"
/>,
            <div>
              <input
                            className="comment-input"
                            onChange={[Function]}
                            onKeyPress={[Function]}
                            value=""
              />
</div>,
          ],
          "className": "right-side",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "comments",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="comment-input"
                onChange={[Function]}
                onKeyPress={[Function]}
                value=""
/>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "comment-input",
                "onChange": [Function],
                "onKeyPress": [Function],
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="outer-image left-side"
>
            <img
                        alt="cat pic"
                        src="/public/assets/images/location"
            />
</div>,
          <div
            className="right-side"
>
            <div
                        className="comments"
            />
            <div>
                        <input
                                    className="comment-input"
                                    onChange={[Function]}
                                    onKeyPress={[Function]}
                                    value=""
                        />
            </div>
</div>,
        ],
        "className": "image-close-up",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="cat pic"
              src="/public/assets/images/location"
/>,
            "className": "outer-image left-side",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "cat pic",
              "src": "/public/assets/images/location",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="comments"
/>,
              <div>
                <input
                                className="comment-input"
                                onChange={[Function]}
                                onKeyPress={[Function]}
                                value=""
                />
</div>,
            ],
            "className": "right-side",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "comments",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <input
                  className="comment-input"
                  onChange={[Function]}
                  onKeyPress={[Function]}
                  value=""
/>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "comment-input",
                  "onChange": [Function],
                  "onKeyPress": [Function],
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
